src/core/Akka.Remote/Transport/Streaming/AsyncQueue.cs

```diff
@@ -0,0 +1,209 @@
+                    }
+                }
+            }
+
+            public void SetItem(T item)
+            {
+                lock (this)
+                {
+                    if (_isCompleted)
+                        return;
+
+                    _value = item;
+                    InternalSetCompleted();
+                }
+            }
+
+            public void SetCanceled()
+            {
+                lock (this)
+                {
+                    if (_isCompleted)
+                        return;
+
+                    _isCanceled = true;
+                    InternalSetCompleted();
+                }
+            }
+
+            private void InternalSetCompleted()
+            {
+                _isCompleted = true;
+
+                if (_continuation != null)
+                    ThreadPool.QueueUserWorkItem(state => ((Action)state).Invoke(), _continuation);
+            }
+
+            public IFutureItem GetAwaiter()
+            {
+                return this;
+            }
+
+            public void GetResult()
+            {
+                lock (this)
+                {
+                    if (!_isCompleted)
+                        throw new InvalidOperationException("Operation is not completed");
+                }
+            }
+
+            public void OnCompleted(Action continuation)
+            {
+                bool alreadyCompleted = false;
+                lock (this)
+                {
+                    _continuation = continuation;
+
+                    if (_isCompleted)
+                        alreadyCompleted = true;
+                }
+
+                if (alreadyCompleted)
+                    ThreadPool.QueueUserWorkItem(state => ((Action)state).Invoke(), continuation);
+            }
+        }
+
+        private readonly Queue<T> _queue;
+        private FutureItem _pendingDequeue;
+
+        private bool _addingCompleted;
+        private bool _isDisposed;
+
+        public AsyncQueue()
+        {
+            _queue = new Queue<T>();
+        }
+
+        public bool Enqueue(T item)
+        {
+            FutureItem completedWaiter = null;
+
+            lock (_queue)
+            {
+                if (_addingCompleted)
+                    return false;
+
+                if (_pendingDequeue != null)
+                {
+                    completedWaiter = _pendingDequeue;
+                    _pendingDequeue = null;
+                }
+                else
+                {
+                    _queue.Enqueue(item);
+                }
+            }
+
+            completedWaiter?.SetItem(item);
+
+            return true;
+        }
+
+        public IFutureItem Dequeue()
+        {
+            FutureItem futureItem = new FutureItem();
+
+            lock (_queue)
+            {
+                if (_pendingDequeue != null)
+                    throw new InvalidOperationException("Dequeue operation is already in progress. This class is single reader.");
```

Comment:
Nitpick: I would explain the intention of the design behind this queue in an XML-DOC comment, for the benefit of future contributors who might need to read and modify this code.